### PR TITLE
Add a corner case for collator-selection

### LINKED ISSUE #360: Add a corner case for increasing/decreasing candidate bond
Specifically, one of the checks should be the following:
1. set low bond
2. register candidate
3. increase bond
4. unregister candidate
5. assert unreserved bond is the original (lower) value

Refer to this: https://github.com/paritytech/cumulus/pull/906/files#diff-3ffb0282ef5c5fd986a407ccac39f35a7d44a293b1719fab3ac28b85ad168e9fR87-R112

### PR DESCRIPTION
**Pull request by Dengjianping**
## Description

closes: #360 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (`manta` or `dolphin`) with right title (start with [Manta] or [Dolphin]),
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests.
- [ ] Updated relevant documentation in the code.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. If this number is updated, then the spec_version must also be updated 
- [ ] If needed, notify the committer this is a draft-release and a tag is needed after merging the PR.
- [ ] Verify benchmarks & weights have been updated for any modified runtime logics
- [ ] If needed, bump `version` for every crate.
- [ ] If import a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- [ ] If needed, update our Javascript/Typescript APIs. These APIs are offcially used by exchanges or community developers.
- [ ] If we're going to issue a new release, freeze the code one week early(it depends, but usually it's one week), ensure we have enough time for related testing.
- [ ] Check if inheriting any upstream runtime storage migrations. If any, perform tests with `try-runtime`.

### CODE REVIEW

File: pallets/collator-selection/src/tests.rs [hunk 1]
```diff
@@ -457,3 +456,63 @@ fn remove_collator_should_work() {
...
+		assert_eq!(0, Balances::reserved_balance(candidate_1));
+		assert_eq!(100, Balances::free_balance(candidate_1));
+
+		// remove candidate_2
+		assert_ok!(CollatorSelection::remove_collator(
+			Origin::signed(RootAccount::get()),
+			candidate_2
+		));
+		// check candidate_2
+		assert_eq!(0, Balances::reserved_balance(candidate_2));
```
Comment (ghzlatarev) [inline pallets/collator-selection/src/tests.rs]: It might be better to reserve some more of the balance in another way in order not check against 0.

**Review by Garandor — APPROVED** [review]
LGTM

File: pallets/collator-selection/src/tests.rs [hunk 2]
```diff
@@ -457,3 +456,63 @@ fn remove_collator_should_work() {
 		);
 	});
 }
+
+#[test]
+fn increase_bond_after_register_candidate() {
+	// It's a corner case:
+	// 1. Set orignal bond as 100KMA.
```
Comment (stechu) [inline pallets/collator-selection/src/tests.rs]: `10` from the code?

File: pallets/collator-selection/src/tests.rs [hunk 3]
```diff
@@ -457,3 +456,63 @@ fn remove_collator_should_work() {
...
+	// 2. Register candidate.
+	// 3. increase bond to 150KMA.
```
Comment (stechu) [inline pallets/collator-selection/src/tests.rs]: `15` from the code?

File: pallets/collator-selection/src/tests.rs [hunk 4]
```diff
@@ -457,3 +456,63 @@ fn remove_collator_should_work() {
...
+	// 3. Unregister candidate.
+	// 4. The owner should get 100KMA back instead of 150KMA.
```
Comment (stechu) [inline pallets/collator-selection/src/tests.rs]: See previous comments.

**Review by stechu — CHANGES_REQUESTED** [review]

File: pallets/collator-selection/src/tests.rs [hunk 5]
```diff
@@ -457,3 +456,63 @@ fn remove_collator_should_work() {
...
+		assert_eq!(new_bond, Balances::reserved_balance(candidate_2));
+
+		// remove candidate_1
+		assert_ok!(CollatorSelection::remove_collator(
+			Origin::signed(RootAccount::get()),
+			candidate_1
+		));
+		// check candidate_1
+		assert_eq!(0, Balances::reserved_balance(candidate_1));
+		assert_eq!(100, Balances::free_balance(candidate_1));
```
Comment (ghzlatarev) [inline pallets/collator-selection/src/tests.rs]: Can you write a comment that they start off with 100 in this mock runtime.

**Review by ghzlatarev — APPROVED** [review]

**Review by stechu — APPROVED** [review]
